### ISSUE TITLE
Add type hints for easier documentation and debugging

### ISSUE THREAD

**Issue by chahak13** [alan-turing-institute/AIrsenal#445]
While working on #393 I realized that maybe introducing type hints in the codebase might be helpful in understanding the code for any contributor and will also help in debugging issues. Furthermore, it can also allow the usage of tools like `mypy` to check and make sure new type bugs don't get introduced further into development. What are your thoughts about this?

**Comment by jack89roberts** [comment]
Yes that would be great, as well as improving docstrings and documentation in general there's a lot we can do to improve the code.
